File: packages/ai/src/normalise.ts
import { flashModel, callWithRetry } from "./client";
import { buildNormalisePrompt } from "./prompts/normalise.prompt";
import { ParsedProfileSchema, ParsedProfile } from "./schemas";

export async function geminiNormalise(rawText: string): Promise<ParsedProfile> {
  // Gemini 2.0 Flash supports up to 1M tokens - can handle full resume text
  const prompt = buildNormalisePrompt(rawText.slice(0, 50000));
  const result = await callWithRetry(() => flashModel.generateContent(prompt));
  const raw = result.response.text();
  return ParsedProfileSchema.parse(JSON.parse(raw));
}

File: frontend/apps/ai/services/normalise.py
import asyncio
from schemas import ParsedProfile
from core.gemini import generate
from prompts.normalise_prompt import build_normalise_prompt


async def gemini_normalise(raw_text: str) -> ParsedProfile:
    """Raw resume text → validated ParsedProfile via Gemini."""
    prompt = build_normalise_prompt(raw_text[:8000])
    loop = asyncio.get_event_loop()

    text = await loop.run_in_executor(None, generate, prompt)
    return ParsedProfile.model_validate_json(text)
